File: src/traccar.js
function cleanBase(url) {
  return url.replace(/\/+$/, "");
}

async function fetchJson(config, path) {
  const headers = config?.token ? { Authorization: `Bearer ${config.token}` } : {};
  const res = await fetch(path, { headers, cache: "no-store" });
  if (!res.ok) throw new Error(`Request failed: ${path}`);
  return res.json();
}

export async function fetchDevices(config) {
  if (!config?.traccarUrl || !config?.token) return [];
  const url = `${cleanBase(config.traccarUrl)}/api/devices`;
  const data = await fetchJson(config, url);
  return Array.isArray(data) ? data : [];
}

export async function fetchPositions(config) {
  if (!config?.traccarUrl || !config?.token) return [];
  const url = `${cleanBase(config.traccarUrl)}/api/positions`;
  const data = await fetchJson(config, url);
  return Array.isArray(data) ? data : [];
}

export async function fetchRecentHistory(config, deviceId, from, to) {
  if (!config?.traccarUrl || !config?.token) return [];
  const params = new URLSearchParams({
    deviceId: String(deviceId),
    from: from.toISOString(),
    to: to.toISOString(),
  });
  const url = `${cleanBase(config.traccarUrl)}/api/reports/route?${params.toString()}`;
  const data = await fetchJson(config, url);
  return Array.isArray(data) ? data : [];
}

File: backend/services/traccar.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Optional

import httpx


class TraccarClient:
    def __init__(self, base_url: Optional[str], token: Optional[str]):
        self.base_url = base_url.rstrip("/") if base_url else None
        self.token = token

    def _headers(self) -> dict:
        if self.token:
            return {"Authorization": f"Bearer {self.token}"}
        return {}

    async def fetch_devices(self) -> list[dict]:
        if not self.base_url or not self.token:
            return []
        url = f"{self.base_url}/api/devices"
        async with httpx.AsyncClient(timeout=15) as client:
            res = await client.get(url, headers=self._headers())
            res.raise_for_status()
            data = res.json()
            return data if isinstance(data, list) else []

    async def fetch_positions(self) -> list[dict]:
        if not self.base_url or not self.token:
            return []
        url = f"{self.base_url}/api/positions"
        async with httpx.AsyncClient(timeout=15) as client:
            res = await client.get(url, headers=self._headers())
            res.raise_for_status()
            data = res.json()
            return data if isinstance(data, list) else []

    async def fetch_route_report(self, device_id: int, from_dt: datetime, to_dt: datetime) -> list[dict]:
        if not self.base_url or not self.token:
            return []
        params = {
            "deviceId": str(device_id),
            "from": from_dt.astimezone(timezone.utc).isoformat(),
            "to": to_dt.astimezone(timezone.utc).isoformat(),
        }
        url = f"{self.base_url}/api/reports/route"
        async with httpx.AsyncClient(timeout=20) as client:
            res = await client.get(url, headers=self._headers(), params=params)
            res.raise_for_status()
            data = res.json()
            return data if isinstance(data, list) else []
